AdyenCard/Utilities/CardType/BinInfoProvider.swift

```diff
@@ -46,59 +48,58 @@ internal final class BinInfoProvider: AnyBinInfoProvider {
+            self.fallbackCardTypeProvider.provide(for: bin,
+                                                  supportedTypes: supportedTypes,
+                                                  completion: completion)
+        }
+
         guard bin.count >= BinInfoProvider.minBinLength else {
-            return fallbackCardTypeProvider.provideInfo(for: bin,
-                                                        supportedTypes: supportedTypes,
-                                                        completion: completion)
+            return fallback()
+        }
+
+        let bin = String(bin.prefix(BinInfoProvider.minBinLength))
+        if let cached = cache[bin] {
+            completion(cached)
+        }
+
+        let useService: (BinLookupService) -> Void = { service in
+            self.use(binLookupService: service, for: bin, supportedTypes: supportedTypes) { [weak self] result in
+                switch result {
+                case let .success(response):
+                    self?.cache[bin] = response
+                    completion(response)
+                case .failure:
+                    fallback()
+                }
+            }
         }
         
-        fetchBinLookupService(
-            success: { [weak self] service in
-                self?.use(binLookupService: service,
-                          for: bin,
-                          supportedTypes: supportedTypes,
-                          completion: completion)
-            },
-            failure: { [weak self] _ in
-                self?.fallbackCardTypeProvider.provideInfo(for: bin, supportedTypes: supportedTypes, completion: completion)
+        if let service = privateBinLookupService {
+            useService(service)
+        } else {
+            fetchBinLookupService { result in
+                switch result {
+                case let .success(service):
+                    useService(service)
+                case .failure:
+                    fallback()
+                }
             }
-        )
+        }
     }
 
     private func use(binLookupService: BinLookupService,
                      for bin: String,
                      supportedTypes: [CardType],
-                     completion: @escaping (BinLookupResponse) -> Void) {
-        binLookupService.requestCardType(for: bin, supportedCardTypes: supportedTypes) { [weak self] result in
-            switch result {
-            case let .success(response):
-                completion(response)
-            case .failure:
-                self?.fallbackCardTypeProvider.provideInfo(for: bin,
-                                                           supportedTypes: supportedTypes,
-                                                           completion: completion)
-            }
-        }
+                     completion: @escaping (Result<BinLookupResponse, Error>) -> Void) {
+        binLookupService.requestCardType(for: bin, supportedCardTypes: supportedTypes, caller: completion)
```

Comment:
now that i look at it after rafactoring, maybe we don't need this function at all?
we can call `binLookupService.requestCardType...` in line 68

Reply:
Indeed 👍